### PR TITLE
New cmdlets for managed database move and copy operations

### PR DESCRIPTION
**Pull request by milanbrkic-ms**
## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by milanbrkic-ms** [comment]
https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1336

File: src/Sql/Sql/ChangeLog.md [hunk 1]
```diff
@@ -19,6 +19,16 @@
 -->
 ## Upcoming Release
 * Added `TryPlannedBeforeForcedFailover` parameter to `Switch-AzSqlDatabaseFailoverGroup`
+* Added new cmdlet for managed database move and copy operations
```
Comment (ostojicb) [inline src/Sql/Sql/ChangeLog.md:22]:
```suggestion
* Added new cmdlets for managed database move and copy operations
```

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetAzureSqlManagedDatabaseCopyOperation.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        DefaultParameterSetName = GetMoveCopyManagedDatabaseOperationsByNameParameterSet),
+        OutputType(typeof(ManagedDatabaseMoveCopyOperation))]
+    public class GetAzureSqlManagedDatabaseCopyOperation : GetMoveCopyAzureSqlManagedDatabaseOperationsBase
+    {
+        protected override OperationMode GetOperationMode()
+        {
+            return OperationMode.COPY;
+        }
+    }
+}
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetAzureSqlManagedDatabaseCopyOperation.cs:30]:
```suggestion
}

```

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/MoveCopyAzureSqlManagedDatabaseBase.cs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+            return sourceDatabase.ToString();
+        }
+
+        protected override string GetConfirmActionProcessMessage()
+        {
+            return ShouldProcessConfirmationMessage;
+        }
+    }
+}
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/MoveCopyAzureSqlManagedDatabaseBase.cs:165]:
```suggestion
}

```

File: src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseCommunicator.cs [hunk 4]
```diff
@@ -178,6 +178,69 @@ public ManagedDatabaseRestoreDetailsResult GetLogReplayStatus(string resourceGro
...
+
+        public IList<ManagedDatabaseMoveOperationResult> GetMoveOperations(string resourceGroupName, string location, string managedInstanceName, string databaseName, string targetManagedInstanceName, OperationMode mode, bool onlyLatestPerDatabase)
+        {
+            var operationMode = mode == OperationMode.MOVE ? "Move" : "Copy";
+            string filter = $"Properties/OperationMode eq '{operationMode}' and (Properties/SourceManagedInstanceName eq '{managedInstanceName}'";
+
+            if (targetManagedInstanceName.Equals(managedInstanceName))
+            {
+                filter += $" or Properties/TargetManagedInstanceName eq '{targetManagedInstanceName}')";
+            }
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseCommunicator.cs:224]: not sure I understand the filtering logic, could you elaborate, why do we have `and` & `or` variants?

File: src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs [hunk 5]
```diff
@@ -261,6 +261,51 @@ public void StopManagedDatabaseLogReplay(AzureSqlManagedDatabaseModel parameters
             }
         }
 
+        public void MoveManagedDatabase(MoveCopyManagedDatabaseModel model)
+        {
+            if (OperationMode.COPY.Equals(model.OperationMode))
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs:266]: you could've just passed the `model.OperationMode` to the API 😄

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseMove.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    {
+        protected override string ShouldProcessConfirmationMessage => "Stop ongoing managed database move operation";
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = "CancelManagedDatabaseMoveByMoveModelObject", HelpMessage = "Object that is returned from start move operation.")]
+        [ValidateNotNullOrEmpty]
+        public MoveCopyManagedDatabaseModel MoveModelObject { get; set; }
+
+        protected override MoveCopyManagedDatabaseModel PersistChanges(MoveCopyManagedDatabaseModel model)
+        {
+            if (MoveModelObject != null)
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseMove.cs:34]: is this expected to ever be null? (with `Mandatory = true` and `ValidateNotNullOrEmpty`)

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseCopy.cs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        protected override string ShouldProcessConfirmationMessage => "Stop ongoing managed database copy operation";
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = "CancelManagedDatabaseCopyByCopyModelObject", HelpMessage = "Object that is returned from start copy operation.")]
+        [ValidateNotNullOrEmpty]
+        public MoveCopyManagedDatabaseModel CopyModelObject { get; set; }
+
+
+        protected override MoveCopyManagedDatabaseModel PersistChanges(MoveCopyManagedDatabaseModel model)
+        {
+            if (CopyModelObject != null)
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseCopy.cs:35]: same question

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                Location = location,
+                OperationMode = GetOperationMode(),
+            };
+
+            return ModelAdapter.ListMoveCopyOperations(model, OnlyLatestPerDatabase.IsPresent);
+        }
+
+
+    }
+}
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:105]:
```suggestion
}

```

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        /// <summary>
+        /// Get the operation mode of the command. COPY or MOVE
+        /// </summary>
+        /// <returns>Operation mode</returns>
+        protected abstract OperationMode GetOperationMode();
+
+
+        protected override IList<ManagedDatabaseMoveCopyOperation> GetEntity()
+        {
+            if (ParameterSetName.Equals(GetMoveCopyManagedDatabaseOperationsByResourceIdParameterSet))
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:78]: do we also need some arg transformation when using `byObject` parameter sets?

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+namespace Microsoft.Azure.Commands.Sql.ManagedDatabase.Cmdlet
+{
+    public abstract class GetMoveCopyAzureSqlManagedDatabaseOperationsBase : AzureSqlManagedDatabaseCmdletBase<IList<ManagedDatabaseMoveCopyOperation>>
+    {
+        protected const string GetMoveCopyManagedDatabaseOperationsByNameParameterSet = "GetMoveCopyManagedDatabaseOperationsByNameParameterSet";
+        protected const string GetMoveCopyManagedDatabaseOperationsByInputObjectParameterSet = "GetMoveCopyManagedDatabaseOperationsByInputObjectParameterSet";
+        protected const string GetMoveCopyManagedDatabaseOperationsByResourceIdParameterSet = "GetMoveCopyManagedDatabaseOperationsByResourceIdParameterSet";
+        protected const string GetMoveCopyManagedDatabaseOperationsByMoveCopyObjectParameterSet = "GetMoveCopyManagedDatabaseOperationsByMoveCopyObjectParameterSet";
+
+        [Parameter(Mandatory = false, HelpMessage = "The name of the instance database.", ParameterSetName = GetMoveCopyManagedDatabaseOperationsByNameParameterSet)]
```
Comment (ostojicb) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:30]: are we okay with no arg being mandatory for example for `GetMoveCopyAzureSqlManagedDatabaseOperationsBase`

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseMove.cs (hunk 6 above)
Comment (ostojicb) [reply src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/StopAzureSqlManagedDatabaseMove.cs:34]: nvm didn't see the base class 😁

File: src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseCommunicator.cs (hunk 4 above)
Comment (milanbrkic-ms) [reply src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseCommunicator.cs:224]: Thanks for catching this. I have copied same implementation from Portal, but actually here we do not want to do it same.

File: src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs (hunk 5 above)
Comment (milanbrkic-ms) [reply src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs:266]: it is an enum, 0 or 1, and I need to pass string "Move", "Copy". Is there a way in c# to do this?

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs (hunk 10 above)
Comment (milanbrkic-ms) [reply src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:30]: InstanceName and ResourceGroupName are required, and they are inherited from base classes.

File: src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs (hunk 5 above)
Comment (ostojicb) [reply src/Sql/Sql/ManagedDatabase/Services/AzureSqlManagedDatabaseAdapter.cs:266]: riight, i think something like `nameof(OperationMode.COPY)` or `OperationMode.COPY.ToString()` should work (just adjust the enum casing if api requires so)

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs (hunk 9 above)
Comment (ostojicb) [reply src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:78]: what if we call it without piping, like 
```
$dbObj123 = Get-AzSqlDb...
Get-AzSqlInstanceDatabaseCopy... -DatabaseObject $dbObj123
```

**Review by vidai-msft — CHANGES_REQUESTED** [review]

File: src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs (hunk 9 above)
Comment (milanbrkic-ms) [reply src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs:78]: You are correct, forgot about this use-case

Comment (milanbrkic-ms) [reply src/Sql/Sql/ManagedDatabase/Cmdlet/MoveCopy/GetMoveCopyAzureSqlManagedDatabaseOperationsBase.cs]: fixed.

**Comment by vidai-msft** [comment]
@milanbrkic-ms Can you please check in all the updated files under Sql.Sdk when generating with autorest. Otherwise, it may have potential issue. If you prefer me to do it, please let me know. Thanks!

**Comment by vidai-msft** [comment]
> @vidai-msft but I have already checked in all files that I changed, and that is one swagger for ManagedDatabaseMoveOperation

That is the reason why we need to make a single commit id. Partial update may have potential issues and make the models/operations not consistent. Let me raise a PR to merge all the updates for SDK that are not related to your swagger.

**Pull request merged**